pkg/updater/resultstore/query.go

```diff
@@ -0,0 +1,334 @@
+	SearchConfiguredTargets(context.Context, *rspb.SearchConfiguredTargetsRequest, ...grpc.CallOption) (*rspb.SearchConfiguredTargetsResponse, error)
+	ExportInvocation(context.Context, *rspb.ExportInvocationRequest, ...grpc.CallOption) (*rspb.ExportInvocationResponse, error)
+}
+
+func isTargetToken(token string) bool {
+	parts := strings.SplitN(token, ":", 2)
+	key := strings.TrimSpace(parts[0])
+
+	if k := strings.ToLower(key); k == "target" || k == "exact_target" {
+		return true
+	}
+	return false
+}
+
+func complexToken(simpleToken string, target bool) (string, error) {
+	logic := map[string]string{
+		"and": "AND",
+		"or":  "OR",
+		"not": "NOT",
+	}
+	parts := strings.SplitN(simpleToken, ":", 2)
+	if len(parts) == 0 {
+		return "", nil
+	}
+	key := strings.TrimSpace(parts[0])
+	if key == "" {
+		return "", nil
+	}
+	var negative bool // true if the token is negative, e.g. "-label:<label>"
+	if strings.HasPrefix(key, "-") {
+		negative = true
+		key = key[1:]
+	}
+	if len(parts) == 1 {
+		// Could be a logical connector...
+		if token, ok := logic[strings.ToLower(key)]; ok {
+			return token, nil
+		}
+		// ...or a parenthesis...
+		if key == "(" || key == ")" {
+			return key, nil
+		}
+		// ...or a bare token, which is disallowed.
+		return "", &QueryError{
+			Reason: Bare,
+			Err:    fmt.Errorf("bare tokens are not allowed: got %s", key),
+		}
+	}
+
+	// Don't allow 'before:', 'after:'...
+	if key == "before" || key == "after" {
+		return "", &QueryError{
+			Reason: BeforeAfter,
+			Err:    errors.New("'before:', 'after:' are not allowed in TestGrid queries"),
+		}
+	}
+	// Ignore 'status' tokens, they aren't useful for TestGrid queries.
+	if key == "status" {
+		return "", nil
+	}
+
+	var token string
+	k := strings.ToLower(key)
+	switch {
+	case k == "user" && target:
+		token = "invocation.invocation_attributes.users:\"%s\""
+	case k == "user":
+		token = "invocation_attributes.users:\"%s\""
+	case k == "label" && target:
+		token = "invocation.invocation_attributes.labels:\"%s\""
+	case k == "label":
+		token = "invocation_attributes.labels:\"%s\""
+	case k == "target" || k == "exact_target":
+		token = "id.target_id=\"%s\""
+	default:
+		// Assume this is a property, ex. "<property_name>:<value>".
+		token = `propertyEquals(` + key + `, "%s")`
+		if target {
+			token = `(invocationPropertyEquals(` + key + `, "%[1]s") OR configurationPropertyEquals(` + key + `, "%[1]s"))`
+		}
+	}
+
+	value := strings.TrimSpace(parts[1])
+	value = strings.Trim(value, `"`)
+	if value == "" {
+		return "", nil
+	}
+	if negative {
+		token = "NOT " + token
+	}
+	return fmt.Sprintf(token, value), nil
+}
+
+var tokenRegex = regexp.MustCompile(`(?i)(?P<token>(?P<key>\S+):(?P<value>".*?"|'.*?'|\S+))|(?P<logic>or|and|not)|(?P<bare>".*?"|'.*?'|\S+)`)
+
+// ConvertReason includes reasons a conversion might fail.
+type ConvertReason int64
+
+const (
```

Comment:
nit: maybe move all const to the beginning of the file? for better readability

Reply:
Done!